[16.0][FIX] partner_multi_company: Keep partner companies aligned

SirPyTech opened the pull request:
**Steps**:
1. In an empty DB with demo data install `l10n_it`, `partner_multi_company`, `stock`

**Actual Behavior**:
```
2025-10-06 10:55:13,769 1 WARNING ? odoo.modules.loading: Module l10n_it demo data failed to install, installed without demo data 
Traceback (most recent call last):
  File "/opt/odoo/custom/src/odoo/odoo/modules/[loading.py](https://loading.py/)", line 91, in load_demo
    load_data(cr, idref, mode, kind='demo', package=package)
  File "/opt/odoo/custom/src/odoo/odoo/modules/[loading.py](https://loading.py/)", line 73, in load_data
    tools.convert_file(cr, package.name, filename, idref, mode, noupdate, kind)
  File "/opt/odoo/custom/src/odoo/odoo/tools/[convert.py](https://convert.py/)", line 771, in convert_file
    convert_xml_import(cr, module, fp, idref, mode, noupdate)
  File "/opt/odoo/custom/src/odoo/odoo/tools/[convert.py](https://convert.py/)", line 837, in convert_xml_import
    obj.parse(doc.getroot())
  File "/opt/odoo/custom/src/odoo/odoo/tools/[convert.py](https://convert.py/)", line 757, in parse
    self._tag_root(de)
  File "/opt/odoo/custom/src/odoo/odoo/tools/[convert.py](https://convert.py/)", line 717, in _tag_root
    raise ParseError(msg) from None  # Restart with "--log-handler odoo.tools.convert:DEBUG" for complete traceback
odoo.tools.convert.ParseError: while parsing /opt/odoo/auto/addons/l10n_it/demo/demo_company.xml:25
The partner must have at least all the companies associated with the user.
View error context:
'-no context-'
```

**Expected behavior**:
Demo data of `l10n_it` is loaded

**Additional context**:
This happens because `stock` creates a new company with admin user (see https://github.com/odoo/odoo/blob/5a991ffaf50d3a12d628bd70e6b2f2c9eb912d84/addons/stock/data/stock_demo.xml#L161), but the field `res.users.company_ids` is only updated at cache level (in https://github.com/odoo/odoo/blob/5a991ffaf50d3a12d628bd70e6b2f2c9eb912d84/odoo/fields.py#L4907) so `res.users.write` is not called so the new company is not added to the partner's companies as it happens when writing to `res.users.company_ids`.

The following `write` on `res.users.company_ids` done by `l10n_it`'s demo data (in https://github.com/odoo/odoo/blob/5a991ffaf50d3a12d628bd70e6b2f2c9eb912d84/addons/l10n_it/demo/demo_company.xml#L27) triggers the validation of the field so the constraint raises a `ValidationErrorì that is downgraded to a WARNING because it's just demo data.

The same happens on write to `res.company.user_ids` but that is way less common because the field is not even shown in the UI.

I believe this is an edge case missed by https://github.com/OCA/multi-company/pull/670 so people involved there might be interested @ArnauCForgeFlow @LoisRForgeFlow @pedrobaeza please let me know what you think.

SirPyTech committed:
[FIX] partner_multi_company: Keep partner companies aligned

When a new company is created with user_ids, the update of the users in done only on cache level so the user's partner's company is not aligned.

pedrobaeza commented on partner_multi_company/tests/test_partner_multi_company.py:
```diff
@@ -166,3 +166,18 @@ def test_can_read_partner_without_company(self):
+    def test_company_creation_with_users(self):
+        admin_user = self.env.ref("base.user_admin")
+        company = self.env["res.company"].create(
+            [
```
Why are you adding a list here whilst it's only one dictionary?

SirPyTech replied:
No specific reason, it's just because `create` expects a list as an argument, I can remove it if you prefer.

pedrobaeza replied:
Not really, `create` admits both a dictionary or a list of dictionary.

SirPyTech replied:
Right, but that's just for backward compatibility so I usually try to avoid it:
```py
    @api.model_create_multi
    @api.returns('self', lambda value: value.id)
    def create(self, vals_list):
        """ create(vals_list) -> records

        Creates new records for the model.

        The new records are initialized using the values from the list of dicts
        ``vals_list``, and if necessary those from :meth:`~.default_get`.

        :param Union[list[dict], dict] vals_list:
            values for the model's fields, as a list of dictionaries::

                [{'field_name': field_value, ...}, ...]

            For backward compatibility, ``vals_list`` may be a dictionary.
            It is treated as a singleton list ``[vals]``, and a single record
            is returned.
[...]
        """
```
(from https://github.com/odoo/odoo/blob/5a991ffaf50d3a12d628bd70e6b2f2c9eb912d84/odoo/models.py#L3925)

I'd prefer to keep it, but as I said I can remove it if you prefer.

pedrobaeza replied:
Yes, I prefer. It's wasting lines for no reason, and such "backward compatibility" (I wouldn't call it that way) is an essential part of the API, and it's not going to be removed. The polymorphism of the `create` method is on purpose.

SirPyTech replied:
No problem, it's done :smile:

SirPyTech committed:
[FIX] partner_multi_company: Propagate user default company

pedrobaeza commented on partner_multi_company/models/res_company.py:
```diff
@@ -0,0 +1,17 @@
+        for vals, company in zip(vals_list, companies):
+            if "user_ids" in vals:
+                for user in company.user_ids:
+                    user.partner_id.company_ids += company
```
No need to over-complicate:

```suggestion
        for company in companies
            for user in company.user_ids:
                    user.partner_id.company_ids += company
```

SirPyTech replied:
:+1: I noticed this in the user's override but not in the company one, thanks! Please check

pedrobaeza commented on partner_multi_company/models/res_company.py:
```diff
@@ -0,0 +1,17 @@
+    _inherit = "res.company"
+
+    @api.model_create_multi
+    def create(self, vals_list):
```
It's also a good habit to document why are you overriding the method:

```suggestion
    def create(self, vals_list):
        # Inject the created company into the allowed companies of users
        # already indicated in the company
```

SirPyTech replied:
Right, done, please check

SirPyTech commented:
I'm still struggling to coordinate all the tests in the repo but reviews could help in the meanwhile

chienandalu commented on partner_multi_company/models/res_users.py:
```diff
@@ -9,10 +9,13 @@ class ResUsers(models.Model):
-        if "company_ids" in vals:
-            res.partner_id.company_ids = vals["company_ids"]
-        return res
+        users = super(ResUsers, self).create(vals)
```
We're stumbling upon some cases where the constraint is raising before being able to ensure the compatibility post actions. So maybe it's useless to perform that check before the following code...

This context would be catched in the constraint, of course :)
```suggestion
        users = super(ResUsers, self.with_context(check_company_id=True)).create(vals)
```

SirPyTech replied:
I'm not a fan of this approach because I've seen how activating/deactivating features with context can end up in a _flags mess_, but I agree it can be a solution, thanks for proposing it!
I would leave it as a last resort when everything else fails.

As for this particular case I have found how to make the tests happy, and I added a test with the use-case that was failing previously, please check!

SirPyTech committed:
[FIX] partner_multi_company: Align partner companies first

write on users triggers a write on the partners that validates the partners' companies, but partners' companies haven't been aligned yet so it raises the partner's companies constraint.
Align the companies in the partners first, so that writing on users (and therefore on partners) will satisfy the constraint.

chienandalu reviewed (approved):
Thanks @SirPyTech

pedrobaeza commented:
/ocabot merge patch

LoisRForgeFlow commented:
@SirPyTech Could you fwport this to v17 and v18?

SirPyTech commented:
> @SirPyTech Could you fwport this to v17 and v18?

Yes we have scheduled a task for it, I'll probably get to it this week or the next one.
